utilities/transactions/transaction_base.cc

```diff
@@ -215,8 +217,16 @@ Status TransactionBaseImpl::Get(const ReadOptions& read_options,
 Status TransactionBaseImpl::GetForUpdate(const ReadOptions& read_options,
                                          ColumnFamilyHandle* column_family,
                                          const Slice& key, std::string* value,
-                                         bool exclusive) {
-  Status s = TryLock(column_family, key, true /* read_only */, exclusive);
+                                         bool exclusive,
+                                         const bool do_validate) {
+  assert(do_validate || read_options.snapshot == nullptr);
```

Comment:
I don't think we should assert user input. Crashing user program (while running in debug mode) when passing invalid argument doesn't feel like something we should do.

Reply:
This is undefined behavior in the API and assert is a way to specify the assumption on the input. We leave the runtime error return just in case there was a corner case that we did not anticipate but during the test we would prefer a highly visible crash over a probably silent error return.
I let @hermanlee , @lth, and @yoshinorim chime in if they prefer it differently.

Reply:
Assuming assert is debug only, then this is fine. I would prefer it return an error on non-debug builds though.

Reply:
Correct. The next line returns the error.

Reply:
We can't assume that RocksDB users never run production with debug build. It has happened and is very likely to be happening somewhere too. Turning user input validation into application crashing will prevent users to do it.

@maysamyabandeh why is it an undefined behavior? The behavior is deterministic: return error to users.

Reply:
I'm not talking about MyRocks, I'm talking about the basic RocksDB users. We do sometimes assert when people do obviously wrong things. For example, we frequently assert iterator is valid while they try to call key() and value(). One reason is that there isn't a status code returned for key() and value() and it doesn't make sense to have just for sanitize user behavior. But in here, there is a status code returned, so I believe user argument validation makese more sense.

Reply:
@siying It is undefined because we decided that it is an invalid input. Removing assert would make it harder for application to find if a corner case in their stress test is violating this assumption. That being said, I think what MyRocks would like see is a good indicative of the application's needs. 
@hermanlee @lth @yoshinorim?

Reply:
Why is it easier for users to find the corner case by crashing the program, than return them an error message telling them what mistake they have made?

Reply:
I let MyRocks team as a user of rocksdb chime in of what behavior would make their life easier.

Reply:
It's not about this concrete case or concrete user. It's about the general principle we follow in this code base. It needs to be all consistent in the code base.

Reply:
The PR is prepared for MyRocks primarily. I think it is fair to hear out their experience as the user before making strong opinions about carved-in-stone principles.

Reply:
We can hear it as an input for revisiting the model of crashing application while users provide invalid arguments, but I won't change opinion in here, even if they say they love it a lot.